Improve "SENTINEL FAILOVER" by using the "FAILOVER" command

gmbnomis opened the pull request:
This PR implements #1291 and consists of the following commits:

Two commits to fix problems with the sentinel tests:

Commit: **Wait for all Sentinels to be connected before starting tests**

Up to now the sentinel test initialization verified that all sentinels detect each other.
However, detection does not imply connection, which led to intermittent failures in the
coordinated failover tests (no leader elected since disconnected sentinels do not take part in a
vote).

Fix this by waiting until no sentinel reports being "disconnected".

Commit: **sentinel-tests: Clean up config after config set tests**

Three commits with the actual implementation:

Commit: **Add option for coordinated failover to Sentinel** (coordinated failover without leader election)
Commit: **Allow Sentinel to recover from a stuck FAILOVER**
Commit: **SENTINEL FAILOVER COORDINATED actually does a leader election** (adds leader election before failover starts)

gmbnomis committed:
Add option for coordinated failover to Sentinel

Valkey supports the "FAILOVER" command to switch primary and replica
roles in a coordinated fashion. Add a "COORDINATED" option to "SENTINEL
FAILOVER". When given, use "FAILOVER" in the Sentinel forced failover
procedure. Since we force the failover, i.e. we don't elect a leader,
the failover must succeed quickly before other Sentinels will regard
the primary as down.

Signed-off-by: Simon Baatz <gmbnomis@gmail.com>

gmbnomis committed:
Allow Sentinel to recover from a stuck FAILOVER

The FAILOVER command may ["encounter some scenarios it can not automatically remediate from and
may get stuck."](https://valkey.io/commands/failover/#failover-abort). The supervision of the
Sentinels is used to handle these situations:

1. "REPLICAOF" is not accepted during a failover. Thus, send a "FAILOVER ABORT" before sending a
   "REPLICAOF" to a node in `sentinelSendReplicaOf()`. (If there is no ongoing failover, the
   resulting error will just be ignored)
2. Sentinel monitors nodes for deviations from the expected state (wrong role or replication).
   Add a check for replicas that are in a failover state for too long and reconfigure them.

In addition, monitoring the failover state is used to find out whether the FAILOVER command
is supported by the Valkey instance. If not, don't use "FAILOVER" or "FAILOVER ABORT".

Signed-off-by: Simon Baatz <gmbnomis@gmail.com>

gmbnomis committed:
sentinel-tests: Clean up config after config set tests

Signed-off-by: Simon Baatz <gmbnomis@gmail.com>

gmbnomis committed:
SENTINEL FAILOVER COORDINATED actually does a leader election

Doing a proper leader election allows to increase the time available
for the actual FAILOVER command, as other Sentinels will not attempt
to initiate another failover during the failover timeout.

Interestingly, Sentinel does not take the reported primary up/down
status into account when counting replies to leader election. Thus,
Sentinel will simply proceed with the failover once we reach the
quorum even if we don't are in O_DOWN state.

Signed-off-by: Simon Baatz <gmbnomis@gmail.com>

gmbnomis committed:
Wait for all Sentinels to be connected before starting tests

Up to now the sentinel test initialization verified that all sentinels detect each other.
However, detection does not imply connection, which led to intermittent failures in the
coordinated failover tests (no leader elected since disconnected sentinels do not take part in a
vote).

Fix this by waiting until no sentinel reports being "disconnected".

Signed-off-by: Simon Baatz <gmbnomis@gmail.com>

gmbnomis commented:
@enjoy-binbin I don't want to be pushy, but this PR is over a month old now. Could you have a look at the PR and/or #1291 when you find some time?

enjoy-binbin commented:
sorry, i dont have enough time to review this, i guess i will be free in the next month. @hwware wen, do you think you have the time to review this in a short time? since you know the sentinel better than me.

hwware commented:
Sure, let me review it. @enjoy-binbin

hwware commented:
> Reference i

@gmbnomis Sorry, let me review it ASAP.

hwware commented:
Could you please provide a CI result which include sentinel test suites, because there is no sentinel test in the commit CI? Thanks

hwware commented on src/commands/sentinel-failover.json:
```diff
@@ -3,22 +3,35 @@
         "function": "sentinelCommand",
+        "history": [
+            [
+                "8.0.0",
```
```suggestion
                "8.1.0",
```

gmbnomis replied:
changed

hwware commented on src/commands/sentinel-failover.json:
```diff
@@ -3,22 +3,35 @@
+                "name": "coordinated",
+                "type": "pure-token",
+                "optional": true,
+                "since": "8.0.0"
```
```suggestion
                "since": "8.1.0"
```

gmbnomis replied:
changed

hwware commented on src/sentinel.c:
```diff
@@ -138,6 +139,14 @@ static mstime_t sentinel_default_failover_timeout = 60 * 3 * 1000;
+/* The failover status of a monitored Valkey instance */
+#define SENTINEL_VALKEY_FAILOVER_NS 0 /* Not supported*/
+#define SENTINEL_VALKEY_NO_FAILOVER 1 /* Failover supported, no failover ongoing */
+#define SENTINEL_VALKEY_FAILOVER 2    /* Failover supported, failover ongoing */
```
The name with VALKEY is weird, it could have better name. Example is SENTINEL_MONITORED_INSTANCE_FAILOVER etc

gmbnomis replied:
Yes, I did not like it much either. I tried to make it different from the SENTINEL_FAILOVER_... constants this way.

I changed it to the names you suggested.

hwware commented on src/sentinel.c:
```diff
@@ -212,6 +221,9 @@ typedef struct sentinelValkeyInstance {
     mstime_t role_reported_time;
     mstime_t replica_conf_change_time; /* Last time replica primary addr changed. */
 
+    int valkey_failover_state;            /* Whether the Valkey instance supports/reports a failover in "master_failover_state" */
```
Comment as above

gmbnomis replied:
changed accordingly

hwware commented on tests/sentinel/tests/18-stuck-failover.tcl:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    wait_for_condition 1000 50 {
+        [string match "*flags master link-pending-commands*" [S 0 SENTINEL MASTER mymaster] ]
```
```suggestion
        [string match "*flags master link-pending-commands*" [S 0 SENTINEL PRIMARY mymaster] ]
```
See if the updated command works

gmbnomis replied:
Thanks, I missed that one. Changed.

hwware commented on tests/sentinel/tests/16-config-set-config-get.tcl:
```diff
@@ -56,3 +56,11 @@ test "SENTINEL CONFIG SET, wrong number of arguments" {
+        assert_equal {OK} [S $id SENTINEL CONFIG SET resolve-hostnames no announce-port 0]
+    }
+    assert_match {*no*0*} [S 1 SENTINEL CONFIG GET resolve-hostnames announce-port]
+    assert_match {announce-port 0} [S 1 SENTINEL CONFIG GET announce-port]
```
Why we need this test? and why we need Line 65 again to get announce-port second time?

gmbnomis replied:
It turns out that the sentinel test framework does not run tests completely independent. Configuration changes like the ones tested in this test are persisted and used in the following tests. 

I considered three options:
1. Change the test framework: While this might be a good idea, I did not want to do this in this PR
2. Undo the config changes at the end of the "config-set-config-get" test.
3. Keep "config-set-config-get" at the end and insert the new tests "manual-coordinated" and "stuck-failover" in the middle, e.g. after "05-manual"

I took option 2 and that's why this code is here. Code lines 64 and 65 just reflect what's done in lines 7 and  8 of the file. You are right that line 65 is redundant in this case and can be deleted.

Alternatively, I could try option 3 and rename the test files.

gmbnomis replied:
I deleted line 65 for the time being.

gmbnomis commented:
> Could you please provide a CI result which include sentinel test suites, because there is no sentinel test in the commit CI? Thanks

[Here](https://github.com/gmbnomis/valkey/actions/runs/12439701547/job/34733941478#step:8:1) is a manual run of the Daily job test-ubuntu-jemalloc with sentinel tests enabled. Is this what you are looking for? If not, please let me know what I should run.

gmbnomis commented:
@hwware Thank you for your review! Given that this PR has been inactive for an additional month, could you please provide guidance on how we can move forward with getting it merged?

hwware commented:
Sorry, I am busying on other stuffs recently. Let me check this PR and https://github.com/valkey-io/valkey/pull/1964 together. I think they are great helpful for sentinel enhancement.

hwware commented:
Please update the json file, version to 9.0 then I will merge this PR first.  I think it takes too long time, Sorry

gmbnomis commented:
> Please update the json file (version to 9.0) then the maintainers have chance to merge this PR

@hwware That's great news. I updated the version number and rebased onto the current `unstable`.
